libs/parallelism/execution/include/hpx/execution/algorithms/bulk.hpp

```diff
@@ -127,28 +124,54 @@ namespace hpx { namespace execution { namespace experimental {
                         });
                 }
             };
-        };
 
-        ///////////////////////////////////////////////////////////////////////
-        template <typename Incrementable>
-        using counting_shape_type = hpx::util::iterator_range<
-            hpx::util::counting_iterator<Incrementable>>;
+            template <typename Receiver>
+            auto connect(Receiver&& receiver) &&
+            {
+                return hpx::execution::experimental::connect(std::move(sender),
+                    bulk_receiver<Receiver>(std::forward<Receiver>(receiver),
+                        std::move(shape), std::move(f)));
+            }
 
-        template <typename Incrementable>
-        HPX_HOST_DEVICE inline counting_shape_type<Incrementable>
-        make_counting_shape(Incrementable n)
-        {
-            return hpx::util::make_iterator_range(
-                hpx::util::make_counting_iterator(Incrementable(0)),
-                hpx::util::make_counting_iterator(n));
-        }
+            template <typename Receiver>
+            auto connect(Receiver&& receiver) &
+            {
+                return hpx::execution::experimental::connect(sender,
+                    bulk_receiver<Receiver>(
+                        std::forward<Receiver>(receiver), shape, f));
+            }
+        };
     }    // namespace detail
 
     ///////////////////////////////////////////////////////////////////////////
     HPX_INLINE_CONSTEXPR_VARIABLE struct bulk_t final
-      : hpx::functional::tag_fallback<bulk_t>
+      : hpx::functional::tag_priority<bulk_t>
     {
     private:
+        // clang-format off
+        template <typename Sender, typename Shape, typename F,
+            HPX_CONCEPT_REQUIRES_(
+                is_sender_v<Sender> &&
+                std::is_integral<Shape>::value &&
+                hpx::execution::experimental::detail::has_completion_scheduler_v<
+                    hpx::execution::experimental::set_value_t, Sender> &&
+                hpx::functional::is_tag_dispatchable_v<bulk_t,
+                    hpx::functional::tag_dispatch_result_t<
+                        hpx::execution::experimental::get_completion_scheduler_t<
+                            hpx::execution::experimental::set_value_t>, Sender>,
+                    Sender, Shape, F>
```

Comment:
Should we combine these two conditions into a separate facility, I would think that can be used in many places.